### PR TITLE
Update reference types support to latest Binaryen

### PR DESCRIPTION
**Pull request by dcodeIO**
Let's see how this goes on CI :)

- [x] I've read the contributing guidelines

### CODE REVIEW

File: tests/compiler/std/set.optimized.wat [hunk 1]
```diff
@@ -4251,14 +4250,57 @@
   i32.store offset=20
   local.get $0
  )
+ (func $~lib/set/Set<u16>#find (param $0 i32) (param $1 i32) (param $2 i32) (result i32)
```
Comment (dcodeIO) [inline tests/compiler/std/set.optimized.wat:4253]: Wondering what happend here. @MaxGraey any ideas why this doesn't optimize away anymore?

File: tests/compiler/std/map.optimized.wat [hunk 2]
```diff
@@ -2559,10 +2546,9 @@
       i32.const 0
      else
       local.get $1
-      i32.load8_u
+      i32.load8_s
       local.get $3
-      i32.const 255
-      i32.and
+      i32.extend8_s
```
Comment (MaxGraey) [inline tests/compiler/std/map.optimized.wat:2551]: That's looks strange little bit. Is it correct changing `load8_u` to `load8_s` and zero-extension to sign-extension?

Comment (dcodeIO) [reply tests/compiler/std/map.optimized.wat:2551]: Hmm, good question. Looks like it's picking the wrong sign extension instruction indeed.

Comment (MaxGraey) [reply tests/compiler/std/map.optimized.wat:2551]: Every unsigned load and every zero-extensions was changed to signed versions. I don't think it's correct. Is it doing by binaryen, right?

Comment (dcodeIO) [reply tests/compiler/std/map.optimized.wat:2551]: Might be a Binaryen optimization doing this, at least didn't find it in the untouched file on a first glimpse. Also, the only place we are emitting `i32.extend8_s` is in `ensureSmallIntegerWrap` on `Type.i8`, which seems correct.

Comment (MaxGraey) [reply tests/compiler/std/map.optimized.wat:2551]: zero-extensions should leave as is. I don't understand why it forced always to sign-extensions

Comment (MaxGraey) [reply tests/compiler/std/map.optimized.wat:2551]: However it doing only for cases like `(x & 0xFFFF) == y` --> `i32.extrend16_s(x) == y` which seems legit due to `i32.eq` / `i32.ne` is sign-agnistic

File: tests/compiler/std/set.optimized.wat (hunk 1 above)
Comment (MaxGraey) [reply tests/compiler/std/set.optimized.wat:4253]: It just add one more monomorphized version for `u16`. I'm wondering could we rewrite `Set#find` in more sign agnostic way?

Comment (MaxGraey) [reply tests/compiler/std/set.optimized.wat:4253]: Ah, I got why this happened. Before `sign-extension` `$~lib/set/Set<u16>#find` and `$~lib/set/Set<i16>#find` was pretty similar and folded (dedup) to one function, but with `sign-extension` they becomes looks pretty different and can't deduplicated anymore

File: tests/compiler/retain-i32.optimized.wat [hunk 3]
```diff
@@ -27,4 +250,7 @@
   i32.load8_s
   drop
  )
+ (func $~start
+  call $start:retain-i32
+ )
```
Comment (MaxGraey) [inline tests/compiler/retain-i32.optimized.wat:255]: `retain-i32.optimized.wat` also pretty huge regression

File: tests/compiler/retain-i32.optimized.wat [hunk 4]
```diff
@@ -1,18 +1,241 @@
...
+ (func $retain-i32/test (param $0 i32) (param $1 i32)
+  (local $2 i32)
+  local.get $0
+  local.get $1
+  i32.add
+  local.tee $2
+  i32.extend8_s
+  local.get $2
+  i32.extend8_s
+  i32.ne
```
Comment (MaxGraey) [inline tests/compiler/retain-i32.optimized.wat:19]: this definitely isolate expressions from pattern matching for `x != x` => `false`. Btw great challenge for improve binaryen in this part.

File: tests/compiler/std/map.optimized.wat (hunk 2 above)
Comment (dcodeIO) [reply tests/compiler/std/map.optimized.wat:2551]: Just to confirm: The optimization here is actually correct, so nothing to worry about?

File: tests/compiler/std/set.optimized.wat (hunk 1 above)
Comment (dcodeIO) [reply tests/compiler/std/set.optimized.wat:4253]: Ah, I see. Makes sense with the signed variant now utilizing sign extension while the unsigned variant does not. Too sad that it also doesn't optimize to the same code anymore.

File: tests/compiler/retain-i32.optimized.wat (hunk 3 above)
Comment (dcodeIO) [reply tests/compiler/retain-i32.optimized.wat:255]: Might this have something to do with sign extension ops not being recognized by precompute or something?

File: tests/compiler/std/map.optimized.wat (hunk 2 above)
Comment (MaxGraey) [reply tests/compiler/std/map.optimized.wat:2551]: yes, it's correct

**Comment by MaxGraey** [comment]
Could you update binaryen again btw?

**Comment by dcodeIO** [comment]
Done, with a few fixture changes.

**Comment by MaxGraey** [comment]
Yes, it was expected. Now binary expressions with float constants also canonicalized always as `x op C`

**Comment by dcodeIO** [comment]
@MaxGraey Are there remaining issues that need to be addressed in here, or is it ok to be merged for now?

**Comment by MaxGraey** [comment]
Could we wait for next nightly binaryen release tonight? It comes with couple optimization enchantments and I want to see how this affect for us

**Comment by dcodeIO** [comment]
Sure! :)

**Comment by MaxGraey** [comment]
Btw I'm not sure we should turn on sign-extention geature by default. Safari still [doesn't support it](https://webassembly.org/roadmap/)

**Comment by dcodeIO** [comment]
Hmm, the matrix looks worrying in regards to when to enable a standardized feature, and from just looking at the table it seems that the difference in features supported by Safari will only get worse.

**Comment by MaxGraey** [comment]
Yes, that' need a discussion.
Btw could you bump binaryen here?

**Comment by dcodeIO** [comment]
Disabled sign extension for now since it's not crucial, but added an agenda item for discussion to the next meeting.

**Comment by dcodeIO** [comment]
Let me know if this looks good to you now :)

**Review by MaxGraey — APPROVED** [review]
LGTM

**Pull request merged**